### PR TITLE
feat : scaled softmax support 

### PR DESCRIPTION
**Pull request by ved1beta**
## feat : scaled softmax support 
#3337 
monkey patch in `scaled_softmax_attn.py`

## Test 
`examples/devstral/devstral-small-qlora.yml`

## Summary by CodeRabbit

* **New Features**
  * Added scaling softmax optimization configuration option for improved attention mechanism performance during training.
  * New configuration parameters: `scaling_softmax` (enable/disable optimization) and `scaling_softmax_factor` (customize scaling behavior).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by winglian** [comment]
do you have loss charts with and without scaled softmax?

**Comment by ved1beta** [comment]
image
with  base flash attention 

with softmax default scaling 1.0 (previous default)

with softmax default scaling 0.5(current default )

**Comment by winglian** [comment]
Do the different values of train loss and perplexity make sense? What should we have expected to see?

**Comment by ved1beta** [comment]
For scaling factor 1 the loss should be identical, which is not probably some math thing fixing!!

**Comment by ved1beta** [comment]
hey here the scale value is learned and should be updated along with weights . currently we are only using fixed value for scale 
image
EDIT: added learned parameter scale for ssm (ready for review ) tested
image

File: src/axolotl/monkeypatch/scaled_softmax_attn.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        Apply SSMax scaling:  score * (s * log(n) + b)
+        where n is the relative position within each packed sequence.
+        """
+        if position_ids_flat is not None:
+            relative_pos = position_ids_flat[q_idx]
+            n = (relative_pos + 1).float()
+        else:
+            n = (q_idx + 1).float()
+
+        n = torch.clamp(n, min=2.0)
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/scaled_softmax_attn.py:91]: what's the motivation behind this? log 1?

File: src/axolotl/utils/schemas/validation.py [hunk 2]
```diff
@@ -201,6 +201,17 @@ def check_sample_packing_with_s2attn(cls, data):
...
         return data
 
+    @model_validator(mode="before")
+    @classmethod
+    def check_scaling_softmax_requires_flex(cls, data):
+        if data.get("scaling_softmax") and not data.get("flex_attention"):
+            raise ValueError(
+                "scaling_softmax requires flex_attention: true\n"
+                "Add 'flex_attention: true' to your config file.\n"
+                "Note: FlexAttention requires PyTorch >= 2.5.0 and Transformers >= 4.48.0"
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/validation.py]:
```suggestion
```
Not needed as our min version exceeds this.

**Review by NanoCode012 — APPROVED** [review]
just minor nit left

**Comment by ved1beta** [comment]
train loss with and without packing with ssm 
image

image_720

**Comment by TheLocalDrummer** [comment]
Awesome! Thanks!

**Comment by TheLocalDrummer** [comment]
Why is 0.43 the scaling factor default? @ved1beta

**Comment by ved1beta** [comment]
paper mentions using 0.43 for figure 1 
image

**Comment by TheLocalDrummer** [comment]
I'm trying this on Ministral 14B Instruct. The difference in loss with and without softmax is quite big. Having a huge loss at the start usually signals incompatibility even if they both eventually plateau near the same loss value.  @ved1beta 

Screenshot 2026-02-02 at 6 27 49 PM

I want to try this on Devstral 123B next, but it's important (to my wallet) that I get promising results with Ministral3 first.

**Comment by TheLocalDrummer** [comment]
Screenshot 2026-02-02 at 6 44 53 PM

Played around with scaling.

Could you share me your Axolotl configs?

**Comment by NanoCode012** [comment]
Hey @TheLocalDrummer , we did not benchmark different values for this hyp. We used the official value in the paper that was mentioned to be good enough default.

I have also sent several message to the mistral team for the hyp value used, but they haven't gotten back on this. Maybe you can nudge them

**Comment by TheLocalDrummer** [comment]
@NanoCode012 If the eval losses for scaling values 0, 0.1, 0.43, and 1.0 indicate anything, it's that this SSM implementation might be broken for Ministral.

**Comment by NanoCode012** [comment]
@TheLocalDrummer , I've nudged mistral again to see if they can share how they did it with details from here.
